packages/medusa/src/services/draft-order.js

```diff
@@ -0,0 +1,433 @@
+      const result = await draftOrderRepo.save(draftOrder)
+
+      await this.eventBus_
+        .withTransaction(manager)
+        .emit(DraftOrderService.Events.CREATED, {
+          id: result.id,
+        })
+
+      let shippingMethods = []
+      let profiles = []
+      if (shippingRequired) {
+        for (const method of shipping_methods) {
+          const m = await this.shippingOptionService_
+            .withTransaction(manager)
+            .createShippingMethod(method.option_id, method.data, {
+              cart: createdCart,
+            })
+
+          shippingMethods.push(m)
+        }
+
+        profiles = shippingMethods.map(
+          ({ shipping_option }) => shipping_option.profile_id
+        )
+      }
+
+      for (const item of items) {
+        if (item.variant_id) {
+          const line = await this.lineItemService_
+            .withTransaction(manager)
+            .generate(
+              item.variant_id,
+              data.region_id,
+              item.quantity,
+              item.metadata
+            )
+
+          const variant = await this.productVariantService_
+            .withTransaction(manager)
+            .retrieve(item.variant_id)
+          const itemProfile = variant.product.profile_id
+
+          let hasShipping = true
+
+          // if shipping is required, ensure items can be shipped
+          if (shippingRequired) {
+            hasShipping = profiles.includes(itemProfile)
+          }
+
+          await this.lineItemService_.withTransaction(manager).create({
+            cart_id: createdCart.id,
+            has_shipping: hasShipping,
+            ...line,
+          })
+        } else {
+          // custom line items can be added to a draft order
+          await this.lineItemService_.withTransaction(manager).create({
+            cart_id: createdCart.id,
+            has_shipping: true,
+            title: item.title || "Custom item",
+            allow_discounts: false,
+            unit_price: item.unit_price || 0,
+            quantity: item.quantity,
+          })
+        }
+      }
+
+      return result
+    })
+  }
+
+  /**
+   * Registers a system payment and creates an order
+   * @param {string} doId - id of draft order to register payment for
+   * @return {Promise<Order>} the created order
+   */
+  async registerSystemPayment(doId) {
+    return this.atomicPhase_(async manager => {
+      const draftOrder = await this.retrieve(doId)
+
+      const draftOrderCart = await this.cartService_
+        .withTransaction(manager)
+        .retrieve(draftOrder.cart_id, {
+          select: ["total"],
+          relations: ["discounts", "shipping_methods", "region", "items"],
+        })
+
+      const orderRepo = manager.getCustomRepository(this.orderRepository_)
+      const draftOrderRepo = manager.getCustomRepository(
+        this.draftOrderRepository_
+      )
+
+      const paymentRepo = manager.getCustomRepository(this.paymentRepository_)
```

Comment:
Similarly I think we should be calling something in `PaymentProviderService` - could be something along the lines of:
```js
const paymentSession = await systemPaymentProvider.create(cart)
const payment = await paymentProviderService.authorizePayment(paymentSession)
await orderService.withTransaction(manager).createFromDraft(draftOrder)
```

Don't think the above is 1:1 with the current interfaces, but the gist of it is really that we ensure that all payments are created through one channel so we don't have to maintain payment creation in multiple places down the road.

Reply:
See above comment.